packages/rabbitmq/src/amqp/connection.ts

```diff
@@ -23,43 +25,133 @@ const defaultConfig = {
   defaultRpcErrorBehavior: MessageHandlerErrorBehavior.REQUEUE,
   defaultSubscribeErrorBehavior: MessageHandlerErrorBehavior.REQUEUE,
   exchanges: [],
-  defaultRpcTimeout: 10000
+  defaultRpcTimeout: 10000,
+  connectionInit: {},
+  connectionManager: {}
 };
 
 export class AmqpConnection {
-  private messageSubject = new Subject<CorrelationMessage>();
-  private _connection!: amqplib.Connection;
-  private _channel!: amqplib.Channel;
-  private config: Required<RabbitMQConfig>;
+  private readonly messageSubject = new Subject<CorrelationMessage>();
+  private readonly config: Required<RabbitMQConfig>;
+  private readonly logger: Logger;
+  private _connection!: amqpcon.AmqpConnectionManager;
+  private _channel!: amqpcon.ChannelWrapper;
+  private _rawChannel?: amqplib.Channel;
+
   constructor(config: RabbitMQConfig) {
     this.config = { ...defaultConfig, ...config };
+    this.logger = new Logger(AmqpConnection.name);
   }
 
   get channel() {
     return this._channel;
   }
 
+  get rawChannel() {
+    return this._rawChannel;
+  }
+
   get connection() {
     return this._connection;
   }
 
-  public async init() {
-    this._connection = await amqplib.connect(this.config.uri);
-    this._channel = await this._connection.createChannel();
-
-    await Promise.all(
-      this.config.exchanges.map(async x =>
-        this._channel.assertExchange(
-          x.name,
-          x.type || this.config.defaultExchangeType,
-          x.options
-        )
+  public async init(): Promise<void> {
+    const {
+      wait = false,
+      timeout = 3000,
+      reject = false
+    } = this.config.connectionInit;
+
+    if (!wait) {
+      return this.initCore();
+    }
+
+    const connectAsync = new Promise(res => this.initCore(() => res()));
+
+    const timeoutAsync = new Promise((res, rej) =>
+      setTimeout(
+        () => (reject ? rej('Failed to connect to a RabbitMQ broker') : res()),
+        timeout
       )
     );
 
-    await this.channel.prefetch(this.config.prefetchCount);
+    return Promise.race<any>([connectAsync, timeoutAsync]);
+  }
+
+  private async initCore(firstInitCallback?: () => void): Promise<void> {
+    this._connection = amqpcon.connect(
+      Array.isArray(this.config.uri) ? this.config.uri : [this.config.uri],
+      this.config.connectionManager
+    );
+
+    this.logger.log('Trying to connect to a RabbitMQ broker');
+    this._connection.on('connect', () => {
+      this.logger.log('Successfully connected to a RabbitMQ broker');
+    });
+
+    this._channel = this._connection.createChannel({
+      name: AmqpConnection.name
+    });
+
+    if (firstInitCallback) {
+      this._channel.once('connect', () => firstInitCallback());
```

Comment:
Is there a specific use case for having a callback for the first connection? It seems like it would be better to have an `onConnection` property which could be `(connectionCount: number) => void` and we could track the number of times a connection has been established as a property of this class. Then a consumer of the library can attach a much more generic callback with more extensible options for processing events

Reply:
I needed that for the purpose of the tests.
Before, the connection module finished bootstrapping after the connection had been made or failed altogether. Therefore, the tests ran after a connection was available and passed successfully.

Now, since the tests had to "wait" for a connection to be made, I made the "init" function that returns a promise and relevant config options to configure it to "wait" mode.
To make it promise-able, I passed down the resolve callback.

I like your approach and we could switch to that and have the tests use it. However, currently this implementation supports the "old" way of having it fail if no connection is available without the developer having to do anything.

I'll add the callback you mentioned, but do you still want to keep this option of wait-and-reject-timeout too?
Also, do you want that property to be a simple callback or would you like to utilize an event-ingn approach of some sort? (EventEmitter / RxJs / etc.)

Reply:
I'll make it simpler. I'll keep the wait+reject mode but I will do that using RxJs observables following your idea.

Reply:
Done